### PR TITLE
[routing-manager] enhance mechanism to publish/unpublish route prefixes

### PR DESCRIPTION
**Pull request by abtink**
This commit updates and enhances the mechanism in `RoutingManager`
to publish/unpublish external route prefixes in Thread Network Data.

Instead of each sub-component (e.g., `DiscoveredPrefixTable`,
`LocalOnLinkPrefix`, `Nat64PrefixManager`, ...) directly trying to
publish/unpublish its entries which can lead to conflicting requests
that then needs to be checked and handled by each component, this
commit adds a new method `EvaluatePublishingPrefix()`. This method
makes a collective decision by checking with different sub-components
to see whether or not each wants this prefix published and if so at
what preference level and flags.

This change help simplify the code and handling of conflicting
situations and edge cases (e.g., while deprecating current/old local
on-link prefix, other routers adding/removing same prefix as PIO or
RIO in their emitted RAs).

-----

Background and info:
- This change centralizes and simplifies the decision making for when to publish/unpublish prefixes in net data
- This allows each sub-component to track its state and avoid the need to check and coordinate with others 
- We have some existing mechanisms in current code to address conflict (by coordinating between components):
   - For example, when we start advertising local on-link prefix we remove the same prefix from `DiscoveredPrefixTable` (without changing the netdata) and then ignore it in any received PIO. 
   - This is mainly to avoid prefix table entry messing up the netdata entry (ubpublish the prefix when we should not).
   - With the change in this PR, we can consider removing this in code later (I kept it as is in this PR but we can remove this in future PRs).
- The existing model does not cover all conflict situations:
   - While deprecating current (or old) local on-link prefix we can receive PIO/RIO adding/removing same prefix which can then mess up the published netdata prefixes.
   -  Or (though unlikely) the same prefix (local on-link prefix) can be advertised (added/removed) as RIO in received RA causing it to be incorrectly unpublished.
   - There may be similar edge cases related to nat64 prefix as well.
- Also the existing model does not scale up as we add more sub-components (more entities that may publish/unpublish)
   - For example, adding mechanism to retain (in non-volatile) and deprecate old prefixes.

This PR contains the commits from two other PRs https://github.com/openthread/openthread/pull/8151 and https://github.com/openthread/openthread/pull/8178 (please check and review the last commit).

### CODE REVIEW

**Review by wgtdkp — APPROVED** [review]
@abtink What about adding unit tests to verify this change covers the described corner cases?

**Comment by abtink** [comment]
> @abtink What about adding unit tests to verify this change covers the described corner cases?

Good idea @wgtdkp. I added a new test case in `test_routing_manager` under `TestConflictingPrefix()`. I checked as I expected this test does indeed fail on `main` (i.e. without the change in this PR).

**Comment by wgtdkp** [comment]
> > @abtink What about adding unit tests to verify this change covers the described corner cases?
> 
> Good idea @wgtdkp. I added a new test case in `test_routing_manager` under `TestConflictingPrefix()`. I checked as I expected this test does indeed fail on `main` (i.e. without the change in this PR).

Thanks @abtink , the new tests LGTM!

File: src/core/border_router/routing_manager.cpp [hunk 1]
```diff
@@ -458,20 +457,53 @@ void RoutingManager::EvaluateOmrPrefix(void)
...
+
+    shouldPublish = mDiscoveredPrefixTable.ShouldPublish(routeConfig);
+
+    if (mLocalOnLinkPrefix.ShouldPublish(routeConfig))
+    {
+        shouldPublish = true;
+    }
+
+#if OPENTHREAD_CONFIG_NAT64_BORDER_ROUTING_ENABLE
+    if (mNat64PrefixManager.ShouldPublish(routeConfig))
```
Comment (jwhui) [inline src/core/border_router/routing_manager.cpp:492]: Each component may independently update `routeConfig`. However, the last one to return will be the one that determines the preference. Is that what we want?

Comment (abtink) [reply src/core/border_router/routing_manager.cpp:492]: Yes. Exactly. The order we check with each component is important.

We don't have this situation today, but we can have a component check previous preference value (set by some other component ahead of it in the list) and decide if it wants to change it or not.

**Review by jwhui — APPROVED** [review]
